src/Microsoft.OpenApi/Services/OpenApiWalker.cs

```diff
@@ -22,81 +24,134 @@ public OpenApiWalker(OpenApiVisitorBase visitor)
         }
 
         /// <summary>
-        /// Walks through the <see cref="OpenApiDocument"/> and validates each element.
+        /// Walks through the <see cref="OpenApiDocument"/> and visits each element.
         /// </summary>
-        /// <param name="doc"></param>
+        /// <param name="doc">OpenApiDocument to be walked</param>
         public void Walk(OpenApiDocument doc)
```

Comment:
That's my concern. Does the customer only walk starting from OpenApiDocument?

Reply:
If you look at the tests, you can Visit an element directly.  If we wanted to expose Walk entry points for sub-graphs we could do that too.  But then in your next comment you suggest that OpenApiValidator should be internal.

Reply:
Calling Visit is not the same as calling Walk though. Visit doesn't recursively walk into the children, right? I think it's still nice to provide a Walk-like feature at every level of the subgraph.

---
In reply to: [158121892](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158121892) [](ancestors = 158121892)

Reply:
Sure it would be nice, but exposing functionality publicly that people can take a dependency on before we have any real experience with how well the validation works seems risky to me.  Why don't we get the validation working as an internal feature and once we are comfortable with it, then consider exposing more.  Unless we have a concrete usecase, I would rather not increase our support surface area.

---
In reply to: [158124882](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158124882) [](ancestors = 158124882,158121892)

Reply:
Do you know of use cases for doing anything else?

---
In reply to: [158115415](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158115415) [](ancestors = 158115415)

Reply:
One possible use case I can envision:

- Define a large reusable component (say, schema) is a separate file and refer to it using Reference. User wants to make sure this component is defined correctly by running a validation on this component (schema), which is technically not in a document.

---
In reply to: [158125996](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158125996) [](ancestors = 158125996,158115415)

Reply:
The only minor problem we have with that at the moment is we don't support reading anything other than a complete document via our public interface.  We will need to do that internally when pulling in external references.

---
In reply to: [158127399](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158127399) [](ancestors = 158127399,158125996,158115415)

Reply:
You are right. This subgraph-level discussion seems to be a recurring theme (writer, reader, and now walker).

I agree with you on this:

"It is worth noting that is much easier to add more granularity in a future release than to take it away."

So I think this level of exposure is fine as is as a beginning point.

On that note, there might be value in adding the Walk methods for all leaf nodes. In case we would like to flip the internal/public switch in the future, it will be a simple string change.

---
In reply to: [158128738](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158128738) [](ancestors = 158128738,158127399,158125996,158115415)